### PR TITLE
[WIP] Barycentric coord / Texture maps for Triangulated surface

### PR DESCRIPTION
**Pull request by dcoeurjo**
# PR Description

WIP: barycentric coord
next step: update the OBJ parser to get the UV coord and text map IO

# Checklist

- [ ] Unit-test of your feature with [Catch](http://dgtal.org/doc/stable/moduleCatch.html).
- [ ] Doxygen documentation of the code completed (classes, methods, types, members...)
- [ ] Documentation module page added or updated.
- [ ] New entry in the [ChangeLog.md](https://github.com/DGtal-team/DGtal/blob/master/ChangeLog.md) added.
- [ ] No warning raised in Debug ```cmake``` mode (otherwise, Travis C.I. will fail).
- [ ] All continuous integration tests pass (Travis & appveyor)

### CODE REVIEW

File: src/DGtal/shapes/MeshTextureHelpers.h [hunk 1]
```diff
@@ -0,0 +1,501 @@
...
+
+      std::vector<UVTriangle> originalVertIndices;
+      std::vector<UVTriangle> originalUVIndices;
+      std::vector<UVTriangle> originalNormalIndices;
+
+      std::ifstream in(filename, std::ios::in);
+      if (!in)
+      {
+        std::cerr << "Cannot open " << filename << std::endl;
+        exit(1);
```
Comment (copyme) [inline src/DGtal/shapes/MeshTextureHelpers.h:163]: That look like a bad idea. Why not to throw an exception?

File: src/DGtal/shapes/MeshTextureHelpers.h [hunk 2]
```diff
@@ -0,0 +1,501 @@
...
+      
+      for (auto v = 0; v < mesh.nbVertices(); v++)
+      {
+        auto n = normal[v];
+        auto t = tan1[v];
+        
+        // Gram-Schmidt orthogonalize
+        tangent[v] = (t - n * n.dot(t)).getNormalized();
+        // Calculate handedness
+     //   tangent[v].w = ((n.crossProduct(t)).dot(tan2[v]) < 0.0) ? -1.0 : 1.0;
```
Comment (copyme) [inline src/DGtal/shapes/MeshTextureHelpers.h:476]: remove?

File: src/DGtal/shapes/MeshTextureHelpers.h [hunk 3]
```diff
@@ -0,0 +1,501 @@
...
+        }
+      }
+      
+      //Building the mesh
+      mesh.build();
+      
+      //Reordering the UV indices per face
+      textMesh = mesh.template makeFaceMap<UVTriangle>();
+      normalMesh = mesh.template makeFaceMap<UVTriangle>();
+      for(auto f=0; f < mesh.nbFaces(); ++f)
```
Comment (copyme) [inline src/DGtal/shapes/MeshTextureHelpers.h:220]: I think no need for auto here, replace with `size_t`?

File: src/DGtal/shapes/MeshTextureHelpers.h [hunk 4]
```diff
@@ -0,0 +1,501 @@
...
+    
+    static
+    Color
+    textureFetch(const TextureImage &tex,
+                 const UV &uvcoord )
+    {
+      auto extent = tex.domain().upperBound() -  tex.domain().lowerBound();
+      Z2i::Point p = { static_cast<Z2i::Point::Coordinate>(std::round(uvcoord[0]*extent[0])),
+                       static_cast<Z2i::Point::Coordinate>(std::round(uvcoord[1]*extent[1]))};
+      //std::cout<<tex(p)<<std::endl;
```
Comment (copyme) [inline src/DGtal/shapes/MeshTextureHelpers.h:317]: remove

**Comment by dcoeurjo** [comment]
and there are few things that should not be in the helper (like the OBJ reader with texcoord...). I'd need to refactor this quick n dirty helper class

File: src/DGtal/shapes/MeshTextureHelpers.h [hunk 5]
```diff
@@ -0,0 +1,501 @@
...
+      pC = C-Pproj;
+      
+      Point va  = (A-B).crossProduct(A-C); // main triangle cross product
+      Point va1 = pB.crossProduct(pC); // p1's triangle cross product
+      Point va2 = pC.crossProduct(pA); // p2's triangle cross product
+      Point va3 = pA.crossProduct(pB); // p3's triangle cross product
+      
+      double  area = va.norm();
+      lambda[0] = (pB.crossProduct(pC)).norm() / area * sgn(va.dot(va1));
+      lambda[1] = (pC.crossProduct(pA)).norm() / area * sgn(va.dot(va2));;
```
Comment (copyme) [inline src/DGtal/shapes/MeshTextureHelpers.h:360]: double `;`

File: src/DGtal/shapes/MeshTextureHelpers.h [hunk 6]
```diff
@@ -0,0 +1,501 @@
...
+      pB = B-Pproj;
+      pC = C-Pproj;
+      
+      Point va  = (A-B).crossProduct(A-C); // main triangle cross product
+      Point va1 = pB.crossProduct(pC); // p1's triangle cross product
+      Point va2 = pC.crossProduct(pA); // p2's triangle cross product
+      Point va3 = pA.crossProduct(pB); // p3's triangle cross product
+      
+      double  area = va.norm();
+      lambda[0] = (pB.crossProduct(pC)).norm() / area * sgn(va.dot(va1));
```
Comment (copyme) [inline src/DGtal/shapes/MeshTextureHelpers.h:359]: `area` is always none zero?

File: src/DGtal/shapes/MeshTextureHelpers.h [hunk 7]
```diff
@@ -0,0 +1,501 @@
...
+                                    const UVMesh &uvMesh,
+                                    const UVMap &uvMap,
+                                    const VertexNormalMesh &normal)
+    {
+  
+      auto tan1 = mesh.template makeVertexMap<RealPoint>({0.,0.,0.});
+      auto tan2 = mesh.template makeVertexMap<RealPoint>({0.,0.,0.});
+      auto tangent = mesh.template makeVertexMap<RealPoint>({0.,0.,0.});
+
+      for(auto faceId=0; faceId < mesh.nbFaces(); ++faceId)
```
Comment (copyme) [inline src/DGtal/shapes/MeshTextureHelpers.h:426]: I do not think so auto is needed here.

File: src/DGtal/shapes/MeshTextureHelpers.h [hunk 8]
```diff
@@ -0,0 +1,501 @@
...
+      auto C = mesh.position(verts[2]); //C
+      
+      UVTriangle uvTriangle = uvMesh[faceId];
+      
+      auto uvA = uvMap[uvTriangle[0]];
+      auto uvB = uvMap[uvTriangle[1]];
+      auto uvC = uvMap[uvTriangle[2]];
+      
+      auto AB = B - A; //(x1,y1,z1)
+      double x1=AB[0], y1=AB[1], z1=AB[2];
```
Comment (copyme) [inline src/DGtal/shapes/MeshTextureHelpers.h:440]: missing spaces before and after `=`.

File: src/DGtal/shapes/MeshTextureHelpers.h [hunk 9]
```diff
@@ -0,0 +1,501 @@
...
+                                    const VertexNormalMesh &normal)
+    {
+  
+      auto tan1 = mesh.template makeVertexMap<RealPoint>({0.,0.,0.});
+      auto tan2 = mesh.template makeVertexMap<RealPoint>({0.,0.,0.});
+      auto tangent = mesh.template makeVertexMap<RealPoint>({0.,0.,0.});
+
+      for(auto faceId=0; faceId < mesh.nbFaces(); ++faceId)
+      {
+      auto verts = mesh.verticesAroundFace(faceId);
```
Comment (copyme) [inline src/DGtal/shapes/MeshTextureHelpers.h:428]: indents are wrong, no?

File: src/DGtal/shapes/MeshTextureHelpers.h [hunk 10]
```diff
@@ -0,0 +1,501 @@
...
+      return tangent;
+    }
+    
+    
+  private:
+    template <typename T>
+    static
+    int sgn(T val)
+    {
+      return (T(0) < val) - (val < T(0));
```
Comment (copyme) [inline src/DGtal/shapes/MeshTextureHelpers.h:488]: I am sure I saw this code on code exchange. Should not you put the link here?

**Comment by copyme** [comment]
Ah, I think it was not ready for a review. Anyways.

**Comment by dcoeurjo** [comment]
I need to close this PR to apply the `bfg` cleanup.
@dcoeurjo  could you please recreate the PR after the release ?
thx

**Pull request closed without merging**